services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/controller/SubnetController.java

```diff
@@ -86,59 +101,106 @@ public SubnetStateJson createSubnetState(@PathVariable String projectid, @Reques
 //            }
 
+            // Synchronous blocking
+            CompletableFuture<Void> allFuture = CompletableFuture.allOf(vpcFuture, macFuture, routeFuture);
+            allFuture.join();
+
+            macResponse = macFuture.join();
+            routeResponse = routeFuture.join();
+            logger.info("Total processing time:" + (System.currentTimeMillis() - start) + "ms");
+
             // set up value of properties for subnetState
-            List<RouteWebObject> routes = subnetState.getRoutes();
+            List<RouteWebObject> routes = inSubnetState.getRoutes();
             if (routes == null) {
                 routes = new ArrayList<>();
             }
             routes.add(routeResponse.getRoute());
-            subnetState.setRoutes(routes);
+            inSubnetState.setRoutes(routes);
             //subnetState.setGatewayIp(ipResponse.getIpState().getIp());
 
+            this.subnetDatabaseService.addSubnet(inSubnetState);
+
+            subnetState = this.subnetDatabaseService.getBySubnetId(inSubnetState.getId());
+            if (subnetState == null) {
+                throw new ResourcePersistenceException();
+            }
+
+            return new SubnetStateJson(subnetState);
+
         } catch (ResourcePersistenceException e) {
+            logger.error(e.getMessage());
             throw new Exception(e);
-        } catch (FallbackException e) {
-            // Route info of subnet rollback
-            List<RouteWebObject> routes = resource.getSubnet().getRoutes();
-            if (routes != null) {
-                RouteWebObject route = routes.get(routes.size() - 1);
-                this.subnetService.routeRollback(route.getId(), resource.getSubnet().getVpcId());
+        } catch (CompletionException e) {
+            routeResponse = (RouteWebJson) routeResponseAtomic.get();
+            macResponse = (MacStateJson) macResponseAtomic.get();
+            logger.error(e.getMessage());
+
+            // Subnet fallback
+            logger.info("subnet fallback start");
+            this.subnetDatabaseService.deleteSubnet(resource.getSubnet().getId());
+            logger.info("subnet fallback end");
+
+            // Route fallback
+            logger.info("Route fallback start");
+            if (routeResponse != null) {
+                RouteWebObject route = routeResponse.getRoute();
+                this.subnetService.routeFallback(route.getId(), resource.getSubnet().getVpcId());
+            }
+            logger.info("Route fallback end");
+
+            // Mac fallback
+            logger.info("Mac fallback start");
+            if (macResponse != null) {
+                this.subnetService.macFallback(macResponse.getMacState().getMacAddress());
+                //this.subnetService.macFallback(UnitTestConfig.macAddress);
             }
+            logger.info("Mac fallback end");
+            throw new Exception(e);
+        } catch (NullPointerException e) {
```

Comment:
What could cause NullPointerException?

Reply:
This is to prevent that when the call step is executed asynchronously, some responses return null, but this repsonse is used later